### PR TITLE
Front end support for sequences

### PR DESCRIPTION
**Pull request by ajreynol**
With this PR, we now support a preliminary draft of a theory of sequences.

This PR adds front end support for sequences in the smt2 parser, in the new API, adds regressions and unit tests for them.

As discussed offline, many of the string kinds are given a sequence version in the external API, but not in the internal one. This means that a special case for such kinds, getKindHelper was added to Term.

We do not yet support proper smt2 printing of sequences, which will require access to this code for Kind conversions (to distinguish e.g. `str.++` from `seq.++`).

### CODE REVIEW

File: src/api/cvc4cppkind.h [hunk 1]
```diff
@@ -2315,10 +2315,153 @@ enum CVC4_PUBLIC Kind : int32_t
...
+   *   -[1]: Term of sort Sequence (sequence s1)
+   *   -[2]: Term of sort Sequence (sequence s2)
+   * Create with:
+   *   mkTerm(Kind kind, Term child1, Term child2)
+   *   mkTerm(Kind kind, const std::vector<Term>& children)
+   */
+  SEQ_PREFIX,
+  /**
+   * Sequence suffix-of.
+   * Checks whether a sequence s1 is a suffix of sequence 2. If sequence s1 is
```
Comment (4tXJ7f) [inline src/api/cvc4cppkind.h:2435]:
```suggestion
   * Checks whether a sequence s1 is a suffix of sequence s2. If sequence s1 is
```

File: src/api/cvc4cppkind.h [hunk 2]
```diff
@@ -2315,10 +2315,153 @@ enum CVC4_PUBLIC Kind : int32_t
...
+   *   -[2]: Term of sort Sequence (subsequence s2)
+   *   -[3]: Term of sort Integer (index i)
+   * Create with:
+   *   mkTerm(Kind kind, Term child1, Term child2, Term child3)
+   *   mkTerm(Kind kind, const std::vector<Term>& children)
+   */
+  SEQ_INDEXOF,
+  /**
+   * Sequence replace.
+   * Replaces a sequence s2 in a sequence s1 with sequence s3. If s2 does not
```
Comment (4tXJ7f) [inline src/api/cvc4cppkind.h:2391]:
```suggestion
   * Replaces the first occurrence of a sequence s2 in a sequence s1 with sequence s3. If s2 does not
```

File: src/api/cvc4cpp.cpp [hunk 3]
```diff
@@ -1507,6 +1563,23 @@ Term Term::getConstArrayBase() const
   return Term(d_solver, d_expr->getConst<ArrayStoreAll>().getExpr());
 }
 
+std::vector<Term> Term::getConstSequenceElements() const
+{
+  CVC4_API_CHECK_NOT_NULL;
+  // CONST_ARRAY kind maps to STORE_ALL internal kind
```
Comment (4tXJ7f) [inline src/api/cvc4cpp.cpp:1569]:
```suggestion
```

File: src/api/cvc4cpp.cpp [hunk 4]
```diff
@@ -1507,6 +1563,23 @@ Term Term::getConstArrayBase() const
...
+  CVC4_API_CHECK(d_expr->getKind() == CVC4::Kind::CONST_SEQUENCE)
+      << "Expecting a CONST_SEQUENCE Term when calling "
+         "getConstSequenceElements()";
+  const std::vector<Node>& elems =
+      d_expr->getConst<ExprSequence>().getSequence().getVec();
+  std::vector<Term> terms;
+  for (const Node& t : elems)
+  {
+    terms.push_back(Term(d_solver, t.toExpr()));
+  }
```
Comment (4tXJ7f) [inline src/api/cvc4cpp.cpp:1579]:
```suggestion
  std::vector<Term> terms = exprVectorToTerms(elems);
```

File: src/parser/smt2/smt2.cpp [hunk 5]
```diff
@@ -156,18 +156,31 @@ void Smt2::addStringOperators() {
...
+    addOperator(api::STRING_LENGTH, "seq.len");
+    addOperator(api::STRING_SUBSTR, "seq.extract");
+    addOperator(api::STRING_CONTAINS, "seq.contains");
+    addOperator(api::STRING_CHARAT, "seq.at");
+    addOperator(api::STRING_INDEXOF, "seq.indexof");
+    addOperator(api::STRING_REPLACE, "seq.replace");
+    addOperator(api::STRING_PREFIX, "seq.prefixof");
+    addOperator(api::STRING_SUFFIX, "seq.suffixof");
+    addOperator(api::STRING_REV, "seq.rev");
+    addOperator(api::STRING_REPLACE_ALL, "seq.replace_all");
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/parser/smt2/smt2.cpp]: Yeah, you're right, and there is no difference in performance.

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/parser/smt2/smt2.cpp]: I've changed this in my later commits.

File: NEWS [hunk 6]
```diff
@@ -4,6 +4,8 @@ Changes since 1.8
 =================
 
 New Features:
+* A new parametric theory of sequences whose syntax is compatible with the
+  syntax for sequences used by z3.
```
Comment (4tXJ7f) [inline NEWS:8]:
```suggestion
  syntax for sequences used by Z3.
```

**Pull request merged**